File: AllPairsShortestPath/AllPairsShortestPath.cpp
#include <fstream>
#include <sstream>
#include <set>
#include <queue>
#include <map>
#include <climits>
#include <vector>
#include <string>
#include "AllPairsShortestPath.h"

using namespace std;

struct AllPairsShortestPath::Compare {
    bool operator() (pair<int, int> const &p1, pair<int, int> const &p2) {
        return p1.first > p2.first;
    }
};

struct AllPairsShortestPath::VertexPair {
    int start;
    int end;
};

AllPairsShortestPath::VertexPair AllPairsShortestPath::parseInput(const string &filePath) {
    string input;
    ifstream file(filePath);
    VertexPair ansPair{};
    if (file.is_open()) {
        getline(file, input, ',');
        ansPair.start = stoi(input.substr(1));
        getline(file, input, ',');
        ansPair.end = stoi(input.substr(1, input.size() - 2));
    }
    return ansPair;
}

vector<string> AllPairsShortestPath::parseGraph() {
    vector<string> edgeStrings;
    ifstream file("./AllPairsShortestPath/graph.txt");
    string input;
    if (file.is_open()) {
        while (getline(file, input, '{')) {
            if (input.length() > 3) {
                edgeStrings.push_back(input.substr(0, input.length() - 2));
            }
        }
        file.close();
    }
    return edgeStrings;
}

int AllPairsShortestPath::dijkstrasArray(const string &filePath) {
    vector<string> edgeStrings = parseGraph();
    VertexPair ansPair = parseInput(filePath);
    int edgeCount = int(edgeStrings.size());
    int edges[edgeCount][3];
    set<int> vertices;
    for (int i = 0; i < edgeCount; i++) {
        string value;
        stringstream ss(edgeStrings[i]);
        getline(ss, value, ',');
        edges[i][0] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][1] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][2] = stoi(value);
    }
    int vertexCount = int(vertices.size());
    int adjMatrix[vertexCount][vertexCount];
    int distanceMins[vertexCount];
    bool finished[vertexCount];
    for (int i = 0; i < vertexCount; i++) {
        distanceMins[i] = INT_MAX;
        finished[i] = false;
        for (int j = 0; j < vertexCount; j++) {
            adjMatrix[i][j] = 0;
        }
    }
    for (int i = 0; i < edgeCount; i++) {
        adjMatrix[edges[i][0] - 1][edges[i][1] - 1] = edges[i][2];
        adjMatrix[edges[i][1] - 1][edges[i][0] - 1] = edges[i][2];
    }
    distanceMins[ansPair.start - 1] = 0;
    for (int _ = 0; _ < vertexCount; _++) {
        int currentIdx = -1;
        int nextMinDistance = INT_MAX;
        for (int i = 0; i < vertexCount; i++) {
            if (distanceMins[i] < nextMinDistance && !finished[i]) {
                nextMinDistance = distanceMins[i];
                currentIdx = i;
            }
        }
        if (currentIdx == -1) {
            throw runtime_error("finished is complete");
        }
        finished[currentIdx] = true;
        for (int i = 0; i < vertexCount; i++) {
            if (!finished[i] && adjMatrix[currentIdx][i]) {
                int costNew = distanceMins[currentIdx] + adjMatrix[currentIdx][i];
                if (costNew < distanceMins[i]) {
                    distanceMins[i] = costNew;
                }
            }
        }
    }
    return distanceMins[ansPair.end - 1];
}

int AllPairsShortestPath::dijkstrasMinHeap(const string &filePath) {
    vector<string> edgeStrings = parseGraph();
    VertexPair ansPair = parseInput(filePath);
    int edgeCount = int(edgeStrings.size());
    int edges[edgeCount][3];
    set<int> vertices;
    for (int i = 0; i < edgeCount; i++) {
        string value;
        stringstream ss(edgeStrings[i]);
        getline(ss, value, ',');
        edges[i][0] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][1] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][2] = stoi(value);
    }
    int vertexCount = int(vertices.size());
    map<pair<int, int>, int> cost;
    int adjList[vertexCount][vertexCount];
    int vertexNeighborCount[vertexCount];
    for (int i = 0; i < vertexCount; i++) {
        vertexNeighborCount[i] = 0;
    }
    for (int i = 0; i < edgeCount; i++) {
        int vertex1Idx = edges[i][0] - 1;
        int vertex2Idx = edges[i][1] - 1;
        int edgeCost = edges[i][2];
        adjList[vertex1Idx][vertexNeighborCount[vertex1Idx]] = vertex2Idx;
        vertexNeighborCount[vertex1Idx]++;
        adjList[vertex2Idx][vertexNeighborCount[vertex2Idx]] = vertex1Idx;
        vertexNeighborCount[vertex2Idx]++;
        cost[pair<int, int>(vertex1Idx, vertex2Idx)] = edgeCost;
        cost[pair<int, int>(vertex2Idx, vertex1Idx)] = edgeCost;
    }
    priority_queue<pair<int, int>, vector<pair<int, int>>, Compare> queue;
    queue.emplace(0, ansPair.start - 1);
    map<int, int> distanceMins;
    for (int i = 0; i < vertexCount; i++) {
        distanceMins[i] = INT_MAX;
    }
    distanceMins[ansPair.start - 1] = 0;
    while (!queue.empty()) {
        pair<int, int> currentNode = queue.top();
        queue.pop();
        for (int i = 0; i < vertexNeighborCount[currentNode.second]; i++) {
            int newVertex = adjList[currentNode.second][i];
            int newCost = currentNode.first + cost[pair<int, int>(currentNode.second, newVertex)];
            if (newCost < distanceMins[newVertex]) {
                distanceMins[newVertex] = newCost;
                queue.emplace(newCost, newVertex);
            }
        }
    }
    return distanceMins[ansPair.end - 1];
}

int AllPairsShortestPath::floydWarshall(const string &filePath) {
    vector<string> edgeStrings = parseGraph();
    int edgeCount = int(edgeStrings.size());
    int edges[edgeCount][3];
    set<int> vertices;
    int totalWeight = 0;
    for (int i = 0; i < edgeCount; i++) {
        string value;
        stringstream ss(edgeStrings[i]);
        getline(ss, value, ',');
        edges[i][0] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][1] = stoi(value);
        vertices.insert(stoi(value));
        getline(ss, value, ',');
        edges[i][2] = stoi(value);
        totalWeight += stoi(value);
    }
    int vertexCount = int(vertices.size());
    int adjMatrix[vertexCount][vertexCount];
    for (int i = 0; i < vertexCount; i++) {
        for (int j = 0; j < vertexCount; j++) {
            adjMatrix[i][j] = totalWeight;
        }
        adjMatrix[i][i] = 0;
    }
    for (int i = 0; i < edgeCount; i++) {
        adjMatrix[edges[i][0] - 1][edges[i][1] - 1] = edges[i][2];
        adjMatrix[edges[i][1] - 1][edges[i][0] - 1] = edges[i][2];
    }
    for (int i = 0; i < vertexCount; i++) {
        for (int j = 0; j < vertexCount; j++) {
            for (int k = j + 1; k < vertexCount; k++) {
                int possible = adjMatrix[j][i] + adjMatrix[i][k];
                if (adjMatrix[j][k] > possible) {
                    adjMatrix[j][k] = possible;
                    adjMatrix[k][j] = possible;
                }
            }
        }
    }
    VertexPair ansPair = parseInput(filePath);
    return adjMatrix[ansPair.start - 1][ansPair.end - 1];
}

File: AllPairsShortestPath/AllPairsShortestPath.py
from collections import defaultdict
import heapq


def dijkstras_array(file_path):
    pair = tuple(int(a) - 1 for a in open(file_path).read()[1:-1].split(', '))
    graph_path = './AllPairsShortestPath/graph.txt'
    edges = [tuple(int(a) for a in edge.split(', ')) for edge in open(graph_path).read()[2:-2].split('}, {')]
    vertices = set(list(map(lambda x: x[0], edges)) + list(map(lambda x: x[1], edges)))
    vertex_count = len(vertices)
    adj_matrix = [[0 for _ in range(vertex_count)] for _ in range(vertex_count)]
    for edge in edges:
        adj_matrix[edge[0] - 1][edge[1] - 1] = edge[2]
        adj_matrix[edge[1] - 1][edge[0] - 1] = edge[2]
    distance_mins = [float('inf') for _ in range(vertex_count)]
    finished = [False for _ in range(vertex_count)]
    distance_mins[pair[0]] = 0
    for _ in range(vertex_count):
        current_idx = -1
        current_cost = float('inf')
        for i, distance in enumerate(distance_mins):
            if distance < current_cost and not finished[i]:
                current_cost = distance
                current_idx = i
        if current_idx == -1:
            return
        finished[current_idx] = True
        for i in range(vertex_count):
            if not finished[i] and adj_matrix[current_idx][i] != 0:
                new_cost = distance_mins[current_idx] + adj_matrix[current_idx][i]
                if new_cost < distance_mins[i]:
                    distance_mins[i] = new_cost
    return distance_mins[pair[1]]


def dijkstras_min_heap(file_path):
    graph_path = './AllPairsShortestPath/graph.txt'
    edges = [tuple(int(a) for a in edge.split(', ')) for edge in open(graph_path).read()[2:-2].split('}, {')]
    pair = tuple(int(a) - 1 for a in open(file_path).read()[1:-1].split(', '))
    adj_list = defaultdict(list)
    cost = {}
    for edge in edges:
        adj_list[edge[0] - 1].append(edge[1] - 1)
        adj_list[edge[1] - 1].append(edge[0] - 1)
        cost[edge[:-1]] = edge[2]
        cost[edge[:-1][::-1]] = edge[2]
    queue = [(0, pair[0])]
    distance_mins = defaultdict(lambda: float('inf'))
    distance_mins[pair[0]] = 0
    while queue:
        current_cost, current_vertex = heapq.heappop(queue)
        for new_vertex in adj_list[current_vertex]:
            new_cost = current_cost + cost[(current_vertex + 1, new_vertex + 1)]
            if new_cost < distance_mins[new_vertex]:
                distance_mins[new_vertex] = new_cost
                heapq.heappush(queue, (new_cost, new_vertex))
    return distance_mins[pair[1]]


def build_data_structure(file_path):
    edges = [tuple(int(a) for a in edge.split(', ')) for edge in open(file_path).read()[2:-2].split('}, {')]
    vertices = set(list(map(lambda x: x[0], edges)) + list(map(lambda x: x[1], edges)))
    vertex_count = len(vertices)
    adj_matrix = [[float('inf') for _ in range(vertex_count)] for _ in range(vertex_count)]
    for edge in edges:
        adj_matrix[edge[0] - 1][edge[1] - 1] = edge[2]
        adj_matrix[edge[1] - 1][edge[0] - 1] = edge[2]
        adj_matrix[edge[0] - 1][edge[0] - 1] = 0
        adj_matrix[edge[1] - 1][edge[1] - 1] = 0
    for i in range(vertex_count):
        print(i)
        for j in range(vertex_count):
            for k in range(j + 1, vertex_count):
                possible = adj_matrix[j][i] + adj_matrix[i][k]
                if adj_matrix[j][k] > possible:
                    adj_matrix[j][k] = possible
                    adj_matrix[k][j] = possible
    return adj_matrix


def floyd_warshall(file_path, data_structure):
    pair = tuple(int(a) - 1 for a in open(file_path).read()[1:-1].split(', '))
    return data_structure[pair[0]][pair[1]]
